src/common/types/row_chunk.cpp

```diff
@@ -0,0 +1,1037 @@
+#include "duckdb/common/types/row_chunk.hpp"
+
+#include "duckdb/common/types/chunk_collection.hpp"
+
+#include <cfloat>
+#include <limits.h>
+
+namespace duckdb {
+
+//! Swaps and encodings are taking from ART
+#define RC_BSWAP16(x) ((uint16_t)((((uint16_t)(x)&0xff00) >> 8) | (((uint16_t)(x)&0x00ff) << 8)))
+
+#define RC_BSWAP32(x)                                                                                                  \
+	((uint32_t)((((uint32_t)(x)&0xff000000) >> 24) | (((uint32_t)(x)&0x00ff0000) >> 8) |                               \
+	            (((uint32_t)(x)&0x0000ff00) << 8) | (((uint32_t)(x)&0x000000ff) << 24)))
+
+#define RC_BSWAP64(x)                                                                                                  \
+	((uint64_t)((((uint64_t)(x)&0xff00000000000000ull) >> 56) | (((uint64_t)(x)&0x00ff000000000000ull) >> 40) |        \
+	            (((uint64_t)(x)&0x0000ff0000000000ull) >> 24) | (((uint64_t)(x)&0x000000ff00000000ull) >> 8) |         \
+	            (((uint64_t)(x)&0x00000000ff000000ull) << 8) | (((uint64_t)(x)&0x0000000000ff0000ull) << 24) |         \
+	            (((uint64_t)(x)&0x000000000000ff00ull) << 40) | (((uint64_t)(x)&0x00000000000000ffull) << 56)))
+
+RowChunk::RowChunk(BufferManager &buffer_manager, idx_t block_capacity, idx_t entry_size)
+    : buffer_manager(buffer_manager), count(0), block_capacity(block_capacity), entry_size(entry_size) {
+	int n = 1;
+	//! little endian if true
+	if (*(char *)&n == 1) {
+		is_little_endian = true;
+	} else {
+		is_little_endian = false;
+	}
+}
+
+RowChunk::RowChunk(RowChunk &other)
+    : buffer_manager(other.buffer_manager), count(0), block_capacity(other.block_capacity),
+      entry_size(other.entry_size), is_little_endian(other.is_little_endian) {
+}
+
+uint8_t RowChunk::FlipSign(uint8_t key_byte) {
+	return key_byte ^ 128;
+}
+
+uint32_t RowChunk::EncodeFloat(float x) {
+	uint64_t buff;
+
+	//! zero
+	if (x == 0) {
+		buff = 0;
+		buff |= (1u << 31);
+		return buff;
+	}
+	//! infinity
+	if (x > FLT_MAX) {
+		return UINT_MAX;
+	}
+	//! -infinity
+	if (x < -FLT_MAX) {
+		return 0;
+	}
+	buff = Load<uint32_t>((const_data_ptr_t)&x);
+	if ((buff & (1u << 31)) == 0) { //! +0 and positive numbers
+		buff |= (1u << 31);
+	} else {          //! negative numbers
+		buff = ~buff; //! complement 1
+	}
+
+	return buff;
+}
+
+uint64_t RowChunk::EncodeDouble(double x) {
```

Comment:
The BSWAPs, FlipSign, EncodeFloat and EncodeDouble can be moved to a common header, which I will do. EncodeData is slightly different for the RowChunk since it does not allocate a unique_ptr<data_t[]>, so I would prefer to keep these separate if that's ok

Reply:
Sounds great!

Reply:
I've thought about this for a bit and we can actually fully avoid code duplication, the functionality of ART and this is the same. I've moved it to a separate header file, and as a bonus, ART now also supports HUGEINT keys.